sycl/doc/extensions/proposed/sycl_ext_oneapi_forward_progress.asciidoc

```diff
@@ -0,0 +1,610 @@
+execution of a kernel requires knowledge of the forward progress guarantees
+of a certain subset of these threads of execution. For example, kernels using
+atomic operations to coordinate sub-groups in the same work-group only require
+the threads of execution associated with sub-groups to provide concurrent or
+parallel forward progress guarantees---there is no requirement for the
+work-item, work-group or root-group threads of execution to provide any
+specific forward progress guarantees.
+
+To facilitate reasoning about work-item coordination, the forward progress
+guarantees of threads of execution in SYCL can be qualified by a _coordination
+scope_. This extension defines the following coordination scopes:
+
+- All work-item threads of execution spawned directly or indirectly from the
+ND-range kernel's initial root thread of execution are considered part of the
+same root-group scope.
+- All work-item threads of execution spawned directly or indirectly from the
+same work-group thread of execution are considered part of the same work-group
+scope.
+- All work-item threads of execution spawned from the same sub-group thread of
+execution are considered part of the same sub-group scope.
+- Each individual work-item thread of execution is the sole member of its own
+work-item scope.
+
+[NOTE]
+====
+Qualifying a thread of execution's forward progress guarantees with a coordination
+scope can be thought of as a shorthand for combining the forward progress
+guarantees provided by each thread of execution in the hierarchy.
+
+For example, consider a hypothetical implementation that creates a separate
+`std::thread` (providing concurrent forward progress guarantees) to execute
+each sub-group, but which permits an unbounded number of work-groups executed
+as tasks (providing parallel forward progress guarantees). With work-group
+coordination scope, each sub-group provides concurrent forward progress
+guarantees: this reflects the guarantees associated with each `std::thread`
+upon creation. With root-group coordination scope, each sub-group provides
+only parallel forward progress guarantees: each sub-group's progress guarantees
+are weakened to reflect that the creation of a sub-group depends upon a
+specific work-group (task) being scheduled and making progress.
+====
+
+[NOTE]
+====
+Choosing a coordination scope is similar to choosing a memory scope for atomic
+operations, in that it describes the potential set of threads of execution
+that a thread _may_ attempt to coordinate with, and must be chosen to reflect
+a kernel's dynamic communication pattern rather than static properties of its
+source code.
+
+For example, consider a kernel using locks/mutexes. Attempting to acquire a
+lock is a blocking operation, and so if two threads of execution providing
+weakly parallel forward progress guarantees attempt to acquire the lock
+simultaneously this will result in a deadlock. To correctly avoid a deadlock,
```

Comment:
Nitpick: is "will result in a deadlock" too strong?

```suggestion
simultaneously this may result in a deadlock. To correctly avoid a deadlock,
```

Reply:
That's true.  Thank you for nit-picking!